commit 091333278df5ed98cd5971cc3736eeb2ac49911e (refs/changes/15/3015/2)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-11-30T21:56:37+00:00 (1 year, 10 months ago)
    
    MANTA-3503 moray pgclient query function creates its own request id

diff --git a/lib/control.js b/lib/control.js
index f7b8b39..22622df 100644
--- a/lib/control.js
+++ b/lib/control.js
@@ -52,7 +52,7 @@ function _getPGHandleAfter(req, cb) {
             if (req.opts.hasOwnProperty('timeout')) {
                 pg.setTimeout(req.opts.timeout);
             }
-
+            pg.setRequestId(req.req_id);
             req.pg = pg;
             cb(null);
         }
@@ -148,7 +148,7 @@ function handlerPipeline(options) {
                     });
                 }
                 dtrace.fire('handler-done', function () {
-                    return [req.msgid, route, handler.name];
+                    return [req.msgid, req.req_id, route, handler.name];
                 });
                 cb(err);
             });
diff --git a/lib/dtrace.js b/lib/dtrace.js
index b3636ce..ffe1d47 100644
--- a/lib/dtrace.js
+++ b/lib/dtrace.js
@@ -91,8 +91,8 @@ var PROBES = {
     // msgid, req_id, route, handler_name, error message
     'handler-error': ['int', 'char *', 'char *', 'char *', 'char *'],
 
-    // msgid, request_type, handler_name
-    'handler-done': ['int', 'char *', 'char *'],
+    // msgid, req_id, request_type, handler_name
+    'handler-done': ['int', 'char *', 'char *', 'char *'],
 
     // msgid, req_id, bucket
     'getbucket-start': ['int', 'char *', 'char *'],
diff --git a/lib/pg.js b/lib/pg.js
index 521bde9..1514d19 100644
--- a/lib/pg.js
+++ b/lib/pg.js
@@ -78,6 +78,7 @@ function PGClient(options) {
     this._defaultTimeout = options.queryTimeout;
     this._queryTimeout = options.queryTimeout;
     this._moray_txn = false;
+    this._reqid = null;
 
     this.log = options.log.child({
         component: 'PGClient',
@@ -113,6 +114,28 @@ PGClient.prototype.setTimeout = function setQueryTimeout(timeout) {
 };
 
 
+/*
+ * Set the request id for this client. This allows us to associate database
+ * queries with a specific RPC being processed by Moray by passing it as an
+ * argument to the 'query-*' dtrace probes.
+ */
+PGClient.prototype.setRequestId = function setRequestId(reqid) {
+    assert.string(reqid, 'reqid');
+    this._reqid = reqid;
+};
+
+
+/*
+ * Called when the postgres client is released back into the pool. Currently,
+ * all new RPCs will overwrite the request id anyway, but this guards against
+ * accidential request id re-use by future users of this client.
+ */
+PGClient.prototype.clearRequestId = function clearRequestId() {
+    assert.string(this._reqid, 'this._reqid');
+    this._reqid = null;
+};
+
+
 /*
  * Restore default timeout in case it was changed, and return this
  * client back to the pool.
@@ -120,6 +143,7 @@ PGClient.prototype.setTimeout = function setQueryTimeout(timeout) {
 PGClient.prototype.release = function clientRelease() {
     assert.equal(false, this._moray_txn, 'finished transaction');
     this.setTimeout(this._defaultTimeout);
+    this.clearRequestId();
     this.pool.release(this);
 };
 
@@ -136,7 +160,11 @@ PGClient.prototype.query = function clientQuery(sql, args) {
     var res = new EventEmitter();
     var self = this;
     var timer;
-    var reqid = libuuid.create();
+
+    // The node-pooling library will periodically invoke this query method on
+    // available PGClients from the pgCheck function. Since there is no RPC
+    // associated with health checks, the _reqid member will be null.
+    var reqid = (this._reqid === null) ? this._reqid : libuuid.create();
     var aborted = false;
 
     function done(event, arg) {
diff --git a/lib/ping.js b/lib/ping.js
index 8407350..22550ef 100644
--- a/lib/ping.js
+++ b/lib/ping.js
@@ -68,7 +68,7 @@ function ping(options) {
                 rpc.fail(startErr);
                 return;
             }
-
+            pg.setRequestId(req_id);
             var req = pg.query('SELECT name FROM buckets_config ' +
                                'LIMIT 1');
 
